### ISSUE TITLE
Setting multiple startup projects doesn't seem to work

### ISSUE THREAD

**Issue by BMBurstein** [Microsoft/PTVS#1341]
I have a solution with 2 flask projects. They both have a similar structure: `runserver.py` at the root of the project and set as startup script, a directory _<app_name>_ which contains `__init.py__` and a few other files. They are set to listen on different ports.  
If I run each project separately they both run fine, even if they are both running simultaneously. If I set them both to be startup projects and press F5, 2 python interpreters are started (two windows), but only the one that is second in the startup order actually runs.

May be related (and maybe not):  
For some reason VS renamed my `__init.py__` files by appending _(<proj_name>)_ to each, but it didn't do the same for any of the other files with the same names, such as `runserver.py` or `utils.py`. When I tried setting a breakpoint in the `runserver.py` file of the one that doesn't seem to run, VS stopped at the same location _but in the file of the other project_!

**Comment by BMBurstein** [comment]
I want to add that each one has it's own virtual env

**Comment by BMBurstein** [comment]
I just tested with a much simpler and easy to reproduce setup (sorry, I should of done this first):

1. Create a new project of type "Blank Flask Web Project"
2. Add another project to the solution of the same type
3. Set them both as startup projects with action "Start"
4. Press F5
5. Bonus: you can see the effect I mentioned by putting a breakpoint in the file that doesn't work, for example on the `if __name__ == '__main__':` line

**Comment by BMBurstein** [comment]
Also, I forgot to mention that I am using VS2015 with the latest updates, as well as the latest PTVS. From the about dialog:

> Microsoft Visual Studio Professional 2015
> Version 14.0.25123.00 Update 2
> Microsoft .NET Framework
> Version 4.6.01055
> 
> Installed Version: Professional

...
> Python Tools for Visual Studio   2.2.40315.00
> Python Tools for Visual Studio provides IntelliSense, projects, templates, Interactive windows, and other support for Python developers.
> 
> Python Tools for Visual Studio - Django Integration   2.2.40315.00
> Provides templates and integration for the Django web framework.
> 
> Python Tools for Visual Studio - Profiling Support   2.2.40315.00
> Profiling support for Python projects.

**Comment by huguesv** [comment]
Hi Baruch,

Thanks for reporting the issues.

Debugger breaking in the wrong project is a known issue: https://github.com/Microsoft/PTVS/issues/176

`__init__.py` being displayed with the folder name in parens is by design, so that you can easily tell them apart when they are opened in the editor.  The name of the files on disk are not affected, they are still `__init__.py`.

I've successfully run multiple startup projects in the past (though I ran in the bug linked above).  I'll try it again and see how that goes.

**Comment by huguesv** [comment]
I created 2 flask projects, each has their own virtual env. I changed the text output so I can identify them and set both to 'Start'.
![image](https://cloud.githubusercontent.com/assets/1696845/16088460/1ae9e4a2-32dc-11e6-80d4-55c68d4507c6.png)

When I hit F5 both python processes started, and web browser opened for each one.
![image](https://cloud.githubusercontent.com/assets/1696845/16088467/24493f7a-32dc-11e6-8415-1944dbe24243.png)

It's possible that the breakpoint bug led you to believe that one of them wasn't starting when it really was starting.  Could you confirm?

**Comment by BMBurstein** [comment]
No, the problem was definitely not just the breakpoint. The first python window didn't display the line that says flask is running on address X.X.X.X:YYYY etc., and when I try to manually enter what should be the correct url into the browser, it gets a browser error the the host was not found.

However, I just tried to reproduce this on my other laptop (VS2015 Community, update 1, PTVS 2.2) and couldn't. I will try updating and try again.

**Comment by BMBurstein** [comment]
BTW, why is the renaming based on project reserved just for `__init__.py`? Why not all duplicate names?

**Comment by BMBurstein** [comment]
This seems to work/fail indeterminately. For my original project (that I asked about in the beginning) it seems to only work the first time I run it after opening the solution anew. any consequent re-run doesn't work. With the flask projects, I created the solution as I described above multiple times. One time, it ran every time. The other times, it ran fine the first time I ran it, but did not work most other runs. This is from one such run, where the browser actually opened for the non-working instance. Usually it didn't even do that:
![image](https://cloud.githubusercontent.com/assets/1149432/16110812/4da6da12-33b8-11e6-85fa-baa9f87f80b1.png)

**Comment by zooba** [comment]
>>> why is the renaming based on project reserved just for  `__init__.py`?

It's actually just changing the display name we provide to VS, so it only applies where VS wants to display a caption rather than use the filename. And we do it for all `__init__.py` files (using the directory name, not the project name) because it's highly likely to be a duplicate. Going through and detecting all duplicate filenames is much more work and would likely have a noticeable performance impact, given that we would need to do the search independently for every single file (i.e. there's no reliable way to do it once and share the results).

>>> This seems to work/fail indeterminately

It's entirely possible that we have a race condition in selecting the port to use:

```
            return Enumerable.Range(new Random().Next(49152, 65536), 60000).Except(
                from connection in IPGlobalProperties.GetIPGlobalProperties().GetActiveTcpConnections()
                select connection.LocalEndPoint.Port
            ).First();
```

If we start two projects simultaneously, then it's very likely the `new Random()` will be seeded identically and both processing will try and start using the same port. This will work for one and fail for the other.

However, looking at the most recent screenshot, we obviously get past this sometimes and hit another issue. Possibly we have some sort of lock in the debugger that doesn't handle this scenario well, so we'll have to create a repro internally so we can debug the debugger.

You've given us enough information to do this already, so thanks! We'll take a look.
